### ISSUE TITLE
Add `#[non_exhaustive]` to all generated structs and enums

### ISSUE THREAD

**Issue by jorgecarleitao** [TethysSvensson/planus#29]
So that new fields can be added to the format in a backward-compatible manner (which is a hallmark of flatbuffers).

**Comment by TethysSvensson** [comment]
I think the main goal here is to make it ergonomic to write "correct" code that handles all variants, including those from future versions of the schema for whatever "correct" means in the setting.

The current approach works by generating an enum with a finite number of cases. When an unknown tag is deserialized, we return an `Err(planus::Error)` with an inner tag of type `ErrorKind::UnknownUnionTag`. This is not ideal in all cases:

* It makes it easy to ignore the error, and propagate it upwards using `?`. This might or might not be what you want depending on the situation.
* The current approach makes it non-ergonomic to have a `catch-all` variant for case added in the future. This means that updating the schema will require updating all most match statements for the union.

However it also comes with a few advantages:

* It makes it very easy to handle all **current** variants.
* It makes it easy to handle unknown tags as errors using the `planus::Error` type (which includes very nice debugging information).

**Alternatives**

I see two things we ways in which we could change the current codegen:

* Add a `#[non_exhaustive]` attribute to the generated enum
* Add an additional `#[non_exhaustive, no_doc] UnknownVariant { variant_tag: u8 }` variant to it, which we would return in place of an error.

We could implement just one of those changes without the other, but I don't think it would make a lot of sense to do so.

**Tradeoffs:**

I think both approaches are very reasonable and which one you prefer will depend on your use-case. For my work, we will prefer the current approach, but I don't think that's necessarily a good reason to default to that one.

I think your suggestion should probably be the default, but with an option to turn it off. Would you be okay with waiting to add this option until we design a configuration file format where you can set the option?

**Comment by TethysSvensson** [comment]
We would probably also want to be able to mark structs as `non_exhaustive`, but that is potentially even less ergonomic.

**Comment by jorgecarleitao** [comment]
I have finished migrating arrow2 to planus and I agree that there are advantages in not using the `non_exhaustive` idiom. In my case in particular, I found out some things that we were not taking into account. So, I think that maybe this is a nice to have as a configuration, but not a strict requirement - i.e. I agree with your assessment. I would say let's close for now and revisit this in a later time when there is
